categories-and-dynamic-typing/dyn-cats.tex

```diff
@@ -0,0 +1,696 @@
+    When can one type represent another? Let's say I have a dynamic
+    type $D$, and I'd like to be able to represent all of the
+    operations of some types $X,Y,Z$.  So for example every function
+    $X\times Y \to Z$ should be implementable as a function
+    $D\times D \to D$, \emph{without} exposing intensional details
+    about how the function is implemented.
+
+    Retracts work very well for this purpose. To see why, let's say we
+    have retractions for $X$ and $Y$ into $D$, that is we have section
+    retraction pairs $e_X : X \to D, p_X : D \to X$ with
+    $p_X \circ e_X = \id_X$ and similarly $e_Y,p_Y$. Then every
+    function $f : X\to Y$ can be encoded as $p_Y \circ f \circ
+    e_X$. Furthermore, we can take \emph{any} function
+    $\phi : D \to D$ and turn it into a function
+    $e_Y \circ \phi \circ p_X : X \to Y$, and if we compose these
+    processes we recover our original $f$:
+
+    \[
+      e_Y \circ (p_Y \circ f \circ e_X) \circ p_X = (e_Y \circ p_Y) \circ f \circ (e_X \circ p_X) = f
+    \]
+
+    Then we can actually \emph{identify} any $f : X \to Y$ with its
+    image $p_Y \circ f \circ e_X : D \to D$, and we get the following
+    natural characterization:
+
+    \[ \text{If } \phi : D \to D, \exists f : X \to Y. p_Y\circ f
+      \circ e_X = \phi \iff (e_Y \circ p_Y) \circ \phi \circ (e_X
+      \circ p_X) = \phi \]
+
+    The interesting thing to note here is that we've reduced the
+    property involving multiple types to just the functions
+    $(e_X\circ p_X),(e_Y\circ p_Y) : D\to D$, which we call
+    $c_X,c_Y$. Since these are functions from $D$ to itself, we can
+    think of them as \emph{untyped} functions. So if $X$ is a retract
+    of $D$, we can describe all functions involving $X$ using $D$ and
+    a single function $c_X$.
+
+    Since $c_X$ arises from a retraction, it has a special property,
+    which is that it is \emph{idempotent}: $c\circ c = c$. The
+    intuition for this property comes from the section-retraction
+    pairs. We think of the idempotent as a \emph{representation} of a
+    retract, identifying some ``good'' subset of a dynamic type by
+    coercing all elements to behave like that type. Then we can
+    identify $X$ with the image of $c : D \to D$.
+  \end{subsection}
+
+  \begin{subsection}{Universal Object: $P_{\omega}$}
+    So now that we've seen that retracts allow us to encode one type
+    in another, we return to our goal, to construct a single type that
+    characterizes some interesting class of types. This is called a
+    \emph{universal type} or in category terms, a \emph{universal
+      object}.
+
+    \begin{definition}
+      A category has a universal object $D$ if every object in the
+      category has a section-retraction pair into $D$.
+    \end{definition}
+
+    One of the remarkable things that Dana Scott accomplished in
+    \cite{scott1976data} was that he constructed a universal
+    countably-based continuous lattice: $P_{\omega}$.
```

Comment:
Footnote idea: since this is HOPL, add pronunciations "D-infinity", "P-omega", etc.

On one hand it's obvious, but then again if you never hear it spoken its hard to know.